### PR TITLE
Support using custom h/template function names

### PR DESCRIPTION
**Pull request by JorgeBucaran**
Hi @substack!

I use Hyperx in my own project and found hyperxify, but couldn't get it to work without some adjustments.

In this PR I've made it so that users consuming hyperx via another library, e.g, choo, hyperapp, etc, can use a pragma directive (like the jsx pragma) with hyperxify.

Also tests and documentation.

To use the feature, users include the pragma directive directly in their source code where they are using a custom, hyperx-based html tagged literal.

```
/** @hyperx h */
```

Now hyperxify will know how to translate

```js
html`<h1>Beep</h1>`
```

into

```js
h('h1', {}, ['Beep'])
```

The user can change the name `h` to whatever they please.

## custom hx/html function

When using a hyperx-based function by requiring the function from another module, hyperxify can't know the name of the function in advance so easily, so by default, I made it to use `html`. 

GitHub's own markdown detects tagged literal functions named `html` and colors them like regular html. So, I think this is a sensitive default.

```js
return html`
    <main>
      <h1>${name}</h1>
    </main>
`
```

This is not an escapable default feature and only comes into effect if the user is requiring a custom hyperx-based tagged literal.

Using the pragma also, the user can use a different name for the html function.

```
/** @hyperx h htm */
```

Now hyperxify will know how to translate

```js
htm`<h1>Bop</h1>`
```

into

```js
h('h1', {}, ['Bop'])
```

// cc @substack @JamesKyburz @yoshuawuyts

### CODE REVIEW

**Comment by substack** [comment]
hyperxify already has a mechanism for detecting the tagged template literal string.

hyperxify looks for the `require('hyperx')` definition:

```
var hx = require('hyperx')
```

and then it looks for when `hx()` is invoked with an argument:

```
var whatever = hx(vdom)
```

and then when somebody uses `whatever` in a tagged template literal, hyperxify will run hyperx on those strings.

I don't think it's a good idea to have two separate mechanisms.

**Comment by jbucaran** [comment]
@substack I noticed that in the source and that has not been tampered with.

The problem arises when the user is consuming a tagged literal function based in hyperx via a 3rd party, like choo that exports yo-yo through, or hyperapp that exports hyperx(h), where h is hyperapp's own h function. The same applies for library using vdom.h, etc.

What happens is that the user has something like this:

```js
var { create, html } = require("my-library-based-in-hyperx")
create(html`<h1>hi</h1>`)
```

Trying to browserify -t hyperxify that above with the current implementation of hyperxify will not do that they user is expecting, in fact, it will do nothing as the tagged literal will not be modified.

Why? Because, just like you've said, hyperxify looks for the require('hyperx') definition, which is nowhere, since the user is consuming hyperx indirectly.

This also fixes an edge case where hyperxify would fail to nullify `require("hyperx")(vdom.h)` in a single line.

**Comment by jbucaran** [comment]
@substack Not trying to push the pragma directive! If you don't want it, I can remove it and still make the PR benefit HyperApp's community (as well as choo's if they manage yo-yoify to work with an `h` function) by having hyperxify use `h` and `html` as the default `h` and `tfn` if none is found, such as is the case when the user is consuming a hyperx based tagged literal via a 3rd party.

**Comment by substack** [comment]
What about exposing more options to configure the presently hard-coded values like the string `'hyperx'` and `h()`? Would that help with your use case? Transforms can be configured with [subarg](https://npmjs.com/package/subarg) syntax in browserify on the command-line or in package.json.

**Comment by jbucaran** [comment]
@substack That could work too, if it's only declaring what the defaults are via CLI or package.json.

All I want is hyperxify to replace `html` for `h(,,,)` in a file where `hyperx` is not used. 

TBH I tried like you're saying too, but could not find an example of how to do that in your handbook here:

* https://github.com/substack/browserify-handbook#writing-your-own

There's a mention of subarg and something along these lines in the section about plugins, but this is a transform right?

Do you have an example of how to do that with transforms?

**Comment by jbucaran** [comment]
Ohhh, I see what you mean, okay.

```
browserify -t [ hyperxify -h h -tfn html ] ...
```

I'll fix this PR and change the title as well.

Can we have defaults too? h->h and tfn->html, that would just make things work out of the box for everyone.

/@substack

**Comment by jbucaran** [comment]
@substack I removed the pragma and hyperxify can be used like this now:

```
browserify -t [ hyperxify --tfn beep --h bop ] main.js > bundle.js
```

Also, by default, if hyperxify can't determine the h and tfn function and none defaults are given, it will use `h` for the hyperscript-style function and `html` for the template function.

**Comment by jbucaran** [comment]
@substack I made the proposed changes, do you think we can merge this?

File: readme.markdown [hunk 1]
```diff
@@ -61,6 +61,31 @@ function render (state) {
...
+beep`<h1>Bop</h1>`
+```
+
+into
+
+```js
+bop('h1', {}, ['Bop'])
+```
+
+* `--tfn`: Name of the template function.
```
Comment (ahdinosaur) [inline readme.markdown]: why `tfn`? maybe `hx`?

Comment (jbucaran) [inline readme.markdown]: Just using the same variable naming as the code [here](https://github.com/substack/hyperxify/blob/master/index.js#L8).

Comment (ahdinosaur) [inline readme.markdown]: ah sweet

Comment (dodekeract) [inline readme.markdown]: What about using [`pragma`](https://www.npmjs.com/package/babel-plugin-transform-react-jsx)?

**Comment by arturi** [comment]
Nice! Could benefit from this too.

File: readme.markdown (hunk 1 above)
Comment (jbucaran) [inline readme.markdown]: @dodekeract https://github.com/substack/hyperxify/pull/8#issuecomment-278526401

**Comment by jbucaran** [comment]
@arturi It'd be nice, but now that hyperx is not a dependency of hyperapp and hyperxify can be used as is without problems, I'm rather ambivalent to this PR :)

**Comment by jbucaran** [comment]
What I'd really like to see merged is this one: https://github.com/substack/hyperx/pull/50.

@substack

**Comment by Rob-pw** [comment]
Referenced from: #10

**Pull request closed without merging**